src/view/subgraph.rs

```diff
@@ -0,0 +1,347 @@
+//! Views into non-hierarchical parts of `PortView`s.
+
+use std::collections::{BTreeSet, HashMap};
+
+use crate::{algorithms::ConvexChecker, Direction, LinkView, NodeIndex, PortIndex, PortView};
+
+use super::filter::PortFiltered;
+
+type NodeCallback = fn(NodeIndex, &SubgraphContext) -> bool;
+type PortCallback = fn(PortIndex, &SubgraphContext) -> bool;
+
+/// View into a subgraph of a PortView.
+///
+/// The subgraph is given by boundary edges that define where one "enters" the
+/// subgraph (incoming boundary edge) and one "leaves" it (outgoing boundary
+/// edges).
+///
+/// A subgraph is well-defined if the incoming/outgoing boundary edges partition
+/// the edges between the children of the root node into three sets:
+///  - boundary edges: either incoming boundary or outgoing boundary edges,
+///  - interior edges: edges such that all the successor edges are either
+///    outgoing boundary edges or interior edges AND all the predecessor edges
+///    are either incoming boundary edges or interior edges,
+///  - exterior edges: edges such that all the successor edges are either
+///    incoming boundary edges or exterior edges AND all the predecessor edges
+///    are either outgoing boundary edges or exterior edges.
+///
+/// Then the subgraph is made of the interior edges and contains all nodes that
+/// are
+///  - adjacent to an interior edge, or
+///  - in-adjacent to an incoming edge and out-adjacent to an outgoing edge.
+///
+/// An intuitive way of looking at this definition is to imagine that the
+/// boundary edges form a wall around the subgraph, and the subgraph is given
+/// by all nodes and edges that can be reached from within without crossing the
+/// wall. The directedness of edges (incoming/outgoing) defines which side of
+/// the wall is inside, and which is outside.
+///
+/// If both incoming and outgoing boundary edges are empty, the subgraph is
+/// taken to be the entire graph.
+///
+/// If an invalid subgraph is defined, then behaviour is undefined.
+///
+/// At initialisation, this performs a one-off expensive computation (linear in
+/// the size of the subgraph) to determine the nodes that are in the subgraph.
+pub type Subgraph<'g, G> = PortFiltered<'g, G, NodeCallback, PortCallback, SubgraphContext>;
+
+/// Internal context used in the [`Subgraph`] adaptor.
+#[derive(Debug, Clone)]
+pub struct SubgraphContext {
+    nodes: HashMap<NodeIndex, bool>,
+    ports: HashMap<PortIndex, bool>,
+    inputs: Vec<PortIndex>,
+    outputs: Vec<PortIndex>,
+}
+
+impl<'a, G: LinkView> Subgraph<'a, G> {
+    /// Create a new subgraph view of `graph`.
+    ///
+    /// ### Arguments
+    ///
+    /// - `graph`: the graph to take a subgraph from,
+    /// - `boundary`: the boundary ports. Target ports are incoming boundary edges,
+    /// and source ports are outgoing boundary edges.
```

Comment:
```suggestion
    /// - `boundary`: the boundary ports. Incoming ports are incoming boundary edges,
    /// and outgoing ports are outgoing boundary edges.
```
(to match the `Direction` variants)